### PR TITLE
Update cloudfront cache, authentication and authorization.

### PR DESCRIPTION
**Pull request by phejimlin**
Please review my PR.

### CODE REVIEW

File: client/dist/auth/app.js [hunk 1]
```diff
@@ -6,9 +6,10 @@ function testToken() {
   // set token to Authorization header
   $.ajax({
     method: 'GET',
-    url: "https://" + location.hostname + '/api/v1/mgnt/test-token',
+    url: "https://r2c5wmub95.execute-api.ap-northeast-1.amazonaws.com/devjim/api/v1/mgnt/users/me",
```
Comment (jonascheng) [inline client/dist/auth/app.js]: Is this for your local testing? This should not be check-in.

File: client/dist/auth/app.js [hunk 2]
```diff
@@ -40,7 +41,8 @@ $(function() {
     var provider = $(event.currentTarget).attr('id');
     $('#token').html('Loading...');
     $('#test-result').html('Loading...');
-    window.location.href = "https://" + location.hostname + '/authentication/signin/' + provider;
+    // window.location.href = "https://" + location.hostname + '/authentication/signin/' + provider;
+    window.location.href = "https://r2c5wmub95.execute-api.ap-northeast-1.amazonaws.com/devjim/authentication/signin/"+provider;
```
Comment (jonascheng) [inline client/dist/auth/app.js]: Same reason above.

File: api/feedback/handler.js [hunk 3]
```diff
@@ -2,49 +2,90 @@
...
+      reject(new Error("400 Bad Request: " + JSON.stringify(event)));
+    } else {
+      user.getOneUser({
+        accountid: event.authAccountid
+      }, function(err, data) {
+        if (err) {
+          reject(err, null);
+        } else {
+          // Authorized: Designer or Admin
+          if (data.accountid === event.authAccountid && data.role === "Admin" || data.role === "Designer"){
```
Comment (jonascheng) [inline api/feedback/handler.js]: Did you verify the logic? data.accountid === event.authAccountid may always failed.

File: api/user/handler.js [hunk 4]
```diff
@@ -8,37 +8,85 @@ module.exports.handler = function(event, context, callback) {
...
+      reject(new Error("400 Bad Request: " + JSON.stringify(event)));
+    } else {
+      user.getOneUser({
+        accountid: event.authAccountid
+      }, function(err, data) {
+        if (err) {
+          reject(err, null);
+        } else {
+          // Authorized: Admin
+          if (data.accountid === event.authAccountid && data.role === "Admin"){
```
Comment (jonascheng) [inline api/user/handler.js]: Did you verify the logic? data.accountid === event.authAccountid may always failed.

File: api/feedback/handler.js [hunk 5]
```diff
@@ -2,49 +2,90 @@
...
+        }
+      });
+    }
+  });
+
+  // A callback handler to decide return is 304 or 200.
+  const cacheCallback = (err, response) => {
+    if (err) {
+      callback(err, response);
+    } else if (event.ifModifiedSince && response.datetime && response.datetime === parseInt(event.ifModifiedSince)) {
```
Comment (jonascheng) [inline api/feedback/handler.js]: Do you intent to implement response.datetime in lower function, such as feedback.js, survey.js, user.js afterward? I don't see any related change in this PR.

File: client/dist/auth/app.js (hunk 1 above)
Comment (phejimlin) [inline client/dist/auth/app.js]: Yes, it should not be check-in. I will fix it.

File: api/feedback/handler.js (hunk 5 above)
Comment (phejimlin) [inline api/feedback/handler.js]: This is for set response header to datetime prevent it cache.

File: api/feedback/handler.js (hunk 3 above)
Comment (phejimlin) [inline api/feedback/handler.js]: Ok, I will fix it.

File: api/user/handler.js (hunk 4 above)
Comment (phejimlin) [inline api/user/handler.js]: OK, I will fix it.

**Pull request merged**